### ISSUE TITLE
Enable registration or passing of a custom triplet loss function

### ISSUE THREAD

**Issue by mihajenko** [beringresearch/ivis#66]
In Python, `Ivis.__init__` accepts a `distance: str` keyword argument, which sets from a dictionary a predefined triplet loss function for that distance metric. Currently, one of the ways to provide a custom distance function is to monkeypatch the `ivis.nn.losses.get_loss_functions`. Other ways to accomplish the same are even messier from the perspectives of usage and implementation.

The nature of dimensionality reduction, especially when dealing with one-hot-encoded categorical features, sometimes requires custom ways to calculate loss. Under the hood, `ivis` has the ability to enable custom loss functions, but any such offerings need to be implemented in a clean and API-idiomatic manner.

---

A custom distance function requires its own triplet loss implementation. `Ivis.__init__` could support an additional keyword argument (e.g. `triplet_loss: Callable[..., ...] = ...`) for users to be able to pass their own.

Alternatively, it could simply be passed inside the existing `distance` kwarg, with its signature changing to `distance: Union[str, Callable[..., ...]]`.

Another way would be to make the losses dictionary built by `ivis.nn.losses.get_loss_functions` a module-level loss function registrar.

Additionally, docs and examples need to be updated on how to correctly implement a custom loss function. With all currently available distance metrics, the triplet loss implementation follows a very similar pattern, and should not be too daunting to attempt to implement.

**Comment by idroz** [comment]
Hi - agree that this would be a useful feature to implement.

Probably a bigger challenge would come in during model saving/loading. A custom loss function would need to have a reliable mechanism of serialisation.

If this is to be implemented, personal preference is to consider one of the latter two approaches that you propose, as they don't introduce additional parameters to `Ivis.__init__`. @Szubie for additional input on this.

If you feel up to it, do fork the repo and we can work jointly on this.

**Comment by mihajenko** [comment]
I want to clarify something: if we define a triplet loss function with a custom distance metric, that means this same metric should also be used when indexing the vectors. Which means custom triplet loss functions are limited by `Annoy`'s metric offerings.

Does this reveal a bug, since the `Annoy` index in this library is set as fixed to `angular`.

For my use case, I have to incorporate the Jaccard distance into the triplet loss function. But the `Annoy` index doesn't have support for it, so my only option is to implement it in `Annoy` as well.

Am I understanding `ivis`' implementation correctly?

I'd be willing to work on this issue, of course. But if the `Annoy` backend poses such fundamental constraints, then we should also define a triplet loss function validation procedure, which should check if the used distance metric are supported on the indexer.

**Comment by Szubie** [comment]
Hi,

I agree that handling the passing of a custom loss function would be a nice feature to have. As you said, the loss functions are written in a quite consistent manner and it shouldn't be too difficult for a user to provide their own in theory. We could provide some guidance in documentation as well if this feature is popular. I would lean towards your suggestion of reusing the existing `distance` keyword argument to keep the API clean.

 
> I want to clarify something: if we define a triplet loss function with a custom distance metric, that means this same metric should also be used when indexing the vectors. Which means custom triplet loss functions are limited by `Annoy`'s metric offerings.
> 
> Does this reveal a bug, since the `Annoy` index in this library is set as fixed to `angular`?

The k-Nearest-Neighbour (KNN) retrieval process using Annoy is currently treated as a completely separate component from the neural network training procedure. We just need to get a KNN matrix out of Annoy, it shouldn't matter whether the metric Annoy uses to retrieve KNN is different from the metric the network uses to embed into the embedding dims. The loss function the user selects only effects how the network learns to embed data in its embedding layers during training, which impacts how the visualizations look. 

We settled on Annoy as a default for ease of use and fast speed. The default of 'angular' distance was used as we found that it performed well across all the datasets we used it on, and did not want to make the API surface too large. If there is interest in selecting different Annoy distance metrics we may provide the option.

It should actually technically be possible to swap out Annoy for another algorithm that retrieves KNN - we are thinking about implementing a feature in ivis where the user can provide an already-built KNN matrix. This would enable users to compute the KNNs using whatever algorithm they want and provide them to ivis.

**Comment by mihajenko** [comment]
I misunderstood the dependency between the KNN metrics and the triplet loss metrics, thank you for clarifying. Apart from possible serialization issues, I don't see much in the way of implementing the custom triplet loss feature.

When someone has a concrete usecase for it, I'll focus on this issue again.

The feature with the bigger priority for me would actually be a swappable KNN index. To that end, I am integrating the [NGT](https://github.com/yahoojapan/NGT) indexer into `ivis`, because it supports the metric that I need. Feature branch is here: https://github.com/mihajenko/ivis/tree/feature/ngt-backend

**Comment by Szubie** [comment]
Hi, since this issue was opened there have been some relevant changes to ivis.

1) It is now possible to pass custom loss functions to ivis as requested in this issue. The simplest way to do this is to pass the desired _callable_ to the `distance` kwarg when constructing the ivis object. (Alternatively you can register a loss function globally using the `register_loss` function in the ivis.nn.losses module and reference by name string).

2) The neighbour retrieval component of ivis is now swappable to any arbitrary implementation by providing an instance of the `collections.abc.Sequence` class to the `neighbour_matrix` kwarg. To implement the methods of the Sequence class, this object must have a `__len__` method that matches the number of rows in dataset, and a `__getitem__` method that returns a list of neighbour indices for the row with the given index. Check the `ivis.data.neighbour_retrieval` module for examples.


Hopefully these changes make it easier to use your own classes to customize the behaviour of ivis.
